staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/types.go

```diff
@@ -67,5 +69,20 @@ type Key struct {
 	Secret string `json:"secret"`
 }
 
+// KMSConfig contains the configuration for the cloudkms service provider to be used for
+// implementing the KMS based encryption transformer, as well as configuration for the
+// KMS transformer.
+type KMSConfig struct {
+	// kind is the name of the cloudkms Service provider to be used.
+	Kind string `json:"kind,omitempty"`
+	// apiVersion is the API version this block has to be parsed as.
+	APIVersion string `json:"apiVersion"`
+	// cacheSize is the maximum number of secrets which are cached in memory. The default value is 1000.
+	// +optional
+	CacheSize int `json:"cacheSize,omitempty"`
+	// config is provider specific configuration and is different for each cloudkms service provider.
+	Config map[string]interface{} `json:"config,omitempty"`
```

Comment:
I tend to think that this should go in the cloud provider initialization, not here.

I am also not convinced that all cloud provider kms implementations will be best configured this way.

I am sorry I didn't see any of the previous PRs in this sequence. Having a giant switch statement inside the generic apiserver library is not a good way to handle this, it is very hard to extend. I have to run now, I'll try to come back to this tomorrow.